OpenRA.Mods.Common/Traits/World/EditorResourceLayer.cs

```diff
@@ -89,57 +126,142 @@ public void UpdateCell(CPos cell)
 					continue;
 
 				UpdateNetWorth(neighbouringTile.Type, neighbouringTile.Density, neighbouringTile.Type, density);
-				neighbouringTile.Density = density;
-				Tiles[neighbouringCell] = neighbouringTile;
+				Tiles[neighbouringCell] = new ResourceLayerContents(neighbouringTile.Type, density);
 
-				CellChanged?.Invoke(neighbouringCell, type);
+				CellChanged?.Invoke(neighbouringCell, neighbouringTile.Type);
 			}
 		}
 
-		void UpdateNetWorth(ResourceType oldType, int oldDensity, ResourceType newType, int newDensity)
+		void UpdateNetWorth(string oldResourceType, int oldDensity, string newResourceType, int newDensity)
 		{
 			// Density + 1 as workaround for fixing ResourceLayer.Harvest as it would be very disruptive to balancing
-			if (oldType != null && oldDensity > 0)
-				NetWorth -= (oldDensity + 1) * oldType.Info.ValuePerUnit;
+			if (oldResourceType != null && oldDensity > 0 && info.DisplayValues.TryGetValue(oldResourceType, out var oldResourceValue))
+				NetWorth -= (oldDensity + 1) * oldResourceValue;
 
-			if (newType != null && newDensity > 0)
-				NetWorth += (newDensity + 1) * newType.Info.ValuePerUnit;
+			if (newResourceType != null && newDensity > 0 && info.DisplayValues.TryGetValue(newResourceType, out var newResourceValue))
+				NetWorth += (newDensity + 1) * newResourceValue;
 		}
 
-		public int CalculateCellDensity(ResourceType type, CPos c)
+		protected virtual int CalculateCellDensity(ResourceLayerContents contents, CPos c)
 		{
 			var resources = Map.Resources;
-			if (type == null || resources[c].Type != type.Info.ResourceType)
+			if (contents.Type == null || !info.Resources.TryGetValue(contents.Type, out var resourceInfo) || resources[c].Type != resourceInfo.ResourceIndex)
 				return 0;
 
+			if (!info.RecalculateResourceDensity)
+				return contents.Density.Clamp(1, resourceInfo.MaxDensity);
+
 			// Set density based on the number of neighboring resources
 			var adjacent = 0;
 			for (var u = -1; u < 2; u++)
 			{
 				for (var v = -1; v < 2; v++)
 				{
 					var cell = c + new CVec(u, v);
-					if (resources.Contains(cell) && resources[cell].Type == type.Info.ResourceType)
+					if (resources.Contains(cell) && resources[cell].Type == resourceInfo.ResourceIndex)
 						adjacent++;
 				}
 			}
 
-			return Math.Max(int2.Lerp(0, type.Info.MaxDensity, adjacent, 9), 1);
+			return Math.Max(int2.Lerp(0, resourceInfo.MaxDensity, adjacent, 9), 1);
+		}
+
+		protected virtual bool AllowResourceAt(string resourceType, CPos cell)
+		{
+			if (!Map.Ramp.Contains(cell) || Map.Ramp[cell] != 0)
+				return false;
+
+			if (!info.Resources.TryGetValue(resourceType, out var resourceInfo))
+				return false;
+
+			// Ignore custom terrain types when spawning resources in the editor
+			var terrainInfo = Map.Rules.TerrainInfo;
+			var terrainType = terrainInfo.TerrainTypes[terrainInfo.GetTerrainInfo(Map.Tiles[cell]).TerrainType].Type;
+
+			// TODO: Check against actors in the EditorActorLayer
```

Comment:
Note that these TODOs are added as documentation for (non-)existing editor limitations, and are not going to be addressed in this PR.